src/mesh/analysis.rs

```diff
@@ -196,28 +197,57 @@ pub fn border_vertex_indices(edge_sharing: &EdgeSharingMap) -> HashSet<u32> {
 /// Finds continuous loops of border edges, starting from a random edge.
 ///
 /// The mesh may contain holes or islands, therefore it may have an unknown
-/// number of edge loops. If two edge loops meet at a single vertex, the result
-/// may be unpredictable and erratic.
-pub fn border_edge_loops(edge_sharing: &EdgeSharingMap) -> Vec<Vec<UnorientedEdge>> {
-    let mut border_edges: Vec<_> = border_edges(edge_sharing).map(UnorientedEdge).collect();
-
-    let mut edge_loops: Vec<Vec<UnorientedEdge>> = Vec::new();
+/// number of edge loops. If only two edge loops meet at a single vertex
+/// (e.g. if the mesh is non-manifold), this border edge chain is not included
+/// in the result.
+pub fn border_edge_loops(edge_sharing: &EdgeSharingMap) -> Vec<Vec<OrientedEdge>> {
+    let mut border_edges: Vec<_> = border_edges(edge_sharing).collect();
+    let mut edge_loops: Vec<Vec<OrientedEdge>> = Vec::new();
 
     while let Some(edge) = border_edges.pop() {
-        let mut current_loop: Vec<UnorientedEdge> = vec![edge];
+        let mut current_chain: VecDeque<OrientedEdge> = VecDeque::with_capacity(1);
+        current_chain.push_back(edge);
+
+        // Try constructing a loop by matching edges with either front or back
+        // of the deque:
+
+        let loop_closed = loop {
+            let mut found = false;
 
-        while current_loop.len() < 3
-            || !current_loop[current_loop.len() - 1].shares_vertex(current_loop[0])
-        {
             for (i, other_edge) in border_edges.iter().enumerate() {
-                if current_loop[current_loop.len() - 1].shares_vertex(*other_edge) {
-                    current_loop.push(border_edges.remove(i));
+                let front = current_chain[0];
+                if other_edge.chains_to(front) {
+                    current_chain.push_front(border_edges.remove(i));
+                    found = true;
+
+                    break;
+                }
+
+                let back = current_chain[current_chain.len() - 1];
+                if back.chains_to(*other_edge) {
+                    current_chain.push_back(border_edges.remove(i));
+                    found = true;
+
                     break;
                 }
             }
-        }
 
-        edge_loops.push(current_loop);
+            // Continue until the loop is closed or until we cannot find a
+            // matching puzzle piece
+            let front = current_chain[0];
+            let back = current_chain[current_chain.len() - 1];
+            let loop_closed = back.chains_to(front);
+
+            if loop_closed || !found {
+                break loop_closed;
+            }
+        };
+
+        // There can theoretically exist chains of edges that do not form border
+        // edge loops, e.g. if the mesh is non-manifold. We do not output those.
```

Comment:
We should treat such cases, though.

Reply:
Could you explain what you mean?

The algorithm treats it. It should be outputting edge *loops* (based on its name), but skips over edge *chains* (or however they should be named). We can alternatively return `Option` or `Result` if the algorithm encounters a non-manifold mesh, where the chains can happen.